### ISSUE TITLE
[question] Why don't awesome-pair-backward-delete delete unbalanced left parentheses?

### ISSUE THREAD

**Issue by BooAA** [manateelazycat/awesome-pair#5]
I make a gif:

![peek 2019-01-29 01-44](https://user-images.githubusercontent.com/32809182/51855265-c8dc5f80-2367-11e9-864d-e6597496836e.gif)

You can see that awesome-pair-backward-delete stop deleting character if encountering unbalanced left parentheses.

I thought this was a bug at first, but after reading the source code I found this might be some feature designed on purpose. For the code below:

```lisp

(defun awesome-pair-backward-delete ()
  (interactive)
  (cond ((awesome-pair-in-string-p)
         (awesome-pair-backward-delete-in-string))
        ((awesome-pair-in-comment-p)
         (backward-delete-char 1))
        ((awesome-pair-after-close-pair-p)
         (awesome-pair-backward-movein-or-delete-close-pair))
        ((awesome-pair-in-empty-pair-p)
         (awesome-pair-backward-delete-in-pair))
        ((not (awesome-pair-after-open-pair-p))
         (backward-delete-char 1))))
```

The last 2 lines seem to be prevent the unbalanced left parentheses from deleting by awesome-pair-backward-delete,  I'm very curious of why you choose to design the function like this? (No offence, just want to know the reason, since I think it's very reasonable to delete unbalanced parentheses if possible)

BTW, thanks for making this useful and well-designed package. I found it fit my need properly without any redundant functionality.

**Comment by manateelazycat** [comment]
Because awesome-pair.el is based on the correct syntax of the current file, when your code has unbalanced parentheses, the syntax of the code has already generated errors.

In the case that the grammatical structure has been wrong, the plugin itself has been unable to speculate on the deletion behavior based on the grammatical structure (although human can distinguish this situation), and if it continues to delete character backward at this time, it is likely to generate more grammatical structural errors.

To put it simply, when the syntax is wrong, awesome-pair.el is very difficult to remove the unbalanced parentheses through parsing.
When the syntax is incorrect, awesome-pair-backward-delete will stop the deletion and let the person fix the grammar problem.

So I recommend using awesome-pair-backward-delete in the case of correct syntax. Backspace is used when the syntax is incorrect.

**Comment by manateelazycat** [comment]
Of course, if you have better code suggestions, please reopen this issue or submit PR, thank you for your feedback.

**Comment by BooAA** [comment]
Thanks for your detail explanation. Will you accept PR for an opposite version of awesome-pair-backward-delete? (maybe we can name it like "awesome-pair-delete")

So by the design above, I think awesome-pair-delete can have behavior like

```lisp
|((foo) ->  |(foo) ;; remove redundant left parentheses
foo|)   ->  foo|)   ;; don't delete single right parentheses if the grammar structure is already wrong

"|\d"   ->   ""  ;; delete escape character at once when cursor before it
"\|d"   ->   ""  ;; delete escape character at once when cursor inside it

|()     ->   (|)   ;; move in when cursor before left parentheses
(|)     ->    |    ;; delete in pair
```
